### ISSUE TITLE
Add to event after the fact

### ISSUE THREAD

**Issue by perholmes** [getsentry/sentry-unity#1257]
Our biggest problem with Sentry is that there's no way for users to give consent to what gets uploaded. The most useful crash is one that contains everything and the kitchen sink, user work, screenshots, logs, but we can't just upload that without asking.

We've requested this feature for years, and have ended up building a system using our own servers. It works, it has a nice UI, but we still think it's very silly that we have to do this. If Sentry wanted to be more concerned about consent and personally identifying information, it would have to adopt a similar system.

Basically, we cannot wholesale upload before asking like Sentry does. We HAVE to do it only after asking the user. If we ask ahead of time, like when installing the app, users panic and the answer is a guaranteed "no".

When a crash happens, we schedule a crash dialog using Sentry's crash ID, and set aside any files that may need to be uploaded after consent has been obtained. Then on the NEXT launch, we present this dialog:

![Crash Dialog](https://user-images.githubusercontent.com/4181396/228566127-c0e154d7-2596-4836-8ea1-fc3946bec741.png)

Here we can ask for everything. If the user opts out completely, we have only their stack trace and breadcrumbs. If they opt in, we have everything. We remember their opt-in status, and especially their email address, because people otherwise only type it the first time.

Upon pressing OK, we upload this to our own servers, and augment the sentry crash with user feedback text, which is the only thing we're allowed to add to a Sentry crash after the fact. In the Sentry crash viewer, it looks like this.

![Comment](https://user-images.githubusercontent.com/4181396/228568113-62f194b9-931f-409c-84dc-f931fa7a5c22.png)

A field further down provides a direct link to the crash uploads in our system. This field was submitted with the original crash, but only points to a valid link if crash data files were subsequently uploaded.

![Screenshot 2023-03-30 002429](https://user-images.githubusercontent.com/4181396/228681438-42a82aeb-2054-4941-91b4-ff5ecda4d40b.png)

In our system, the crash upload is unzipped, and the files are accessible.

![Viewer](https://user-images.githubusercontent.com/4181396/228681468-6d13fd21-3884-453f-84ce-f052454d39ef.png)

FEATURE REQUEST SUMMARY:

* Sentry should allow augmenting a crash after the fact by uploading attachments, logs, screenshots and fields to ADD to an existing crash, which allows the user to give consent.
* Sentry should have easy ways to view these attachments. We format the logs as plain text or JSON depending on the file. We'll add an image viewer as well.

Consider that Sentry encourages a perverse incentive of uploading as MUCH as possible ahead of time, while being morally and legally obligated to upload as LITTLE as possible without asking consent. It just doesn't fly.

There's no reason we should have to side-load an entire second crash viewer just so we can ask the user permission first. **This entire problem can be solved by allowing subsequent uploads to an existing crash.**

**Comment by bitsandfoxes** [comment]
Hey @perholmes, thanks for the feedback!

The update of already ingested events is something that does not work as of right now. The way Sentry is built simply does not allow for that and most likely won't do so in the foreseeable future.

Sentry takes the concern regarding PII very seriously which is why no PII gets sent by default.

In regards to the feature request of not sending events until the user has given their consent: This is something the `sentry-native` already does and that is considered to be implemented in the other SDKs as well. We're tracking the progress over here https://github.com/getsentry/team-mobile/issues/24 but there is no clear timeline yet.

Allowing to add attachments in i.e. `beforeSend` is something we implemented in other SDKs as `hints` already. AFAIK this is something that is currently being worked on in the .NET SDK and should make its way to the Unity SDK shortly after.

**Comment by mattjohnsonpint** [comment]
Thanks for the excellent feedback @perholmes. We are tracking improvements for a user-consent workflow in various SDKs already.  For Unity, I'd keep an eye on #1044, and the subtasks linked there.

As @bitsandfoxes mentioned, we're also working on adding support for `hints` to the Sentry .NET SDK.  When done, Unity would get this feature also.  See https://github.com/getsentry/sentry-dotnet/issues/1469

Thanks.

**Comment by bitsandfoxes** [comment]
Thanks again for your detailed feedback and feature request. I'm closing this issue in favor of https://github.com/getsentry/sentry-unity/issues/1044
Please feel free to add additional thoughts or concerns you might have.
